src/poetry/puzzle/provider.py

```diff
@@ -947,6 +940,22 @@ def debug(self, message: str, depth: int = 0) -> None:
 
             self._io.write(debug_info)
 
+    def _group_by_source(
```

Comment:
I just realized we don't have a docstring here -- could you add one describing the intention of this method and how it's meant to be used?

The battle against no source code level documentation depends on each PR :smile: